optuna/integration/sklearn.py

```diff
@@ -94,6 +99,48 @@ def _check_sklearn_availability():
         )
 
 
+# NOTE Original implementation:
+# https://github.com/scikit-learn/scikit-learn/blob/ \
+# 8caa93889f85254fc3ca84caa0a24a1640eebdd1/sklearn/utils/validation.py#L131-L135
+def _is_arraylike(x):
+    # type: (Any) -> bool
+
+    return (
+        hasattr(x, '__len__') or
+        hasattr(x, 'shape') or
+        hasattr(x, '__array__')
+    )
+
+
+# NOTE Original implementation:
+# https://github.com/scikit-learn/scikit-learn/blob/ \
+# 8caa93889f85254fc3ca84caa0a24a1640eebdd1/sklearn/utils/validation.py#L217-L234
+# It removed the check if an input is scipy sparse matrix
+def _make_indexable(iterable):
+    # type: (Iterable) -> (Iterable)
+    if hasattr(iterable, "__getitem__") or hasattr(iterable, "iloc"):
+        return iterable
+    elif iterable is None:
+        return iterable
+    return np.array(iterable)
+
+
+def _num_samples(x):
+    # type: (ArrayLikeType) -> int
+
+    # NOTE For dask dataframes
+    # https://github.com/scikit-learn/scikit-learn/blob/ \
+    # 8caa93889f85254fc3ca84caa0a24a1640eebdd1/sklearn/utils/validation.py#L155-L158
+    if hasattr(x, 'shape') and x.shape is not None:
```

Comment:
mypy throws the following errors. (https://app.circleci.com/jobs/github/optuna/optuna/29090)

```
#!/bin/bash -eo pipefail
. venv/bin/activate
mypy --disallow-untyped-defs --ignore-missing-imports .
optuna/integration/sklearn.py:134: error: Item "List[Any]" of "Union[List[Any], Any, Any]" has no attribute "shape"
optuna/integration/sklearn.py:135: error: Item "List[Any]" of "Union[List[Any], Any, Any]" has no attribute "shape"
optuna/integration/sklearn.py:136: error: Item "List[Any]" of "Union[List[Any], Any, Any]" has no attribute "shape"
optuna/integration/sklearn.py:136: error: Incompatible return value type (got "Integral", expected "int")
Found 4 errors in 1 file (checked 121 source files)

Exited with code exit status 1
```

`ArrayLikeType` is defined [here](https://github.com/himkt/optuna/blob/c568c64c8a7c5ed64020341cce018e36ea1fa5fd/optuna/integration/sklearn.py#L52) and I don't understand why these errors occur. 🤕
(`ArrayLikeType = Union[List, np.ndarray, pd.Series]`)

Reply:
I think `mypy` cannot infer the type based on `hasattr`.
How about using `getattr`? It is suggested [here](https://github.com/python/mypy/issues/1424#issuecomment-223828130).

Example:

```python
    x_shape = getattr(x, 'shape', None)
    if x_shape is not None:
        if isinstance(x_shape[0], Integral):
            return int(x_shape[0])
```

Reply:
It works. Thank you!